adapters/metax/metax.go

```
@@ -0,0 +1,244 @@
+	}
+
+	var metaxExt openrtb_ext.ExtImpMetaX
+	if err := json.Unmarshal(bidderExt.Bidder, &metaxExt); err != nil {
+		return nil, errors.New("Wrong MetaX bidder ext")
+	}
+
+	return &metaxExt, nil
+}
+
+// publisher ID and adunit is numeric at this moment
+func validateParams(ext *openrtb_ext.ExtImpMetaX) error {
+	if _, err := strconv.Atoi(ext.PublisherID); err != nil {
+		return errors.New("invalid publisher ID")
+	}
+	if _, err := strconv.Atoi(ext.Adunit); err != nil {
+		return errors.New("invalid adunit")
+	}
+	return nil
+}
+
+func preprocessImp(imp *openrtb2.Imp) error {
+	if imp == nil {
+		return errors.New("imp is nil")
+	}
+
+	if imp.Banner != nil {
+		bannerCopy, err := assignBannerSize(imp.Banner)
+		if err != nil {
+			return err
+		}
+		imp.Banner = bannerCopy
+	}
+
+	// clean inventory
+	switch {
+	case imp.Video != nil:
+		imp.Banner = nil
+		imp.Native = nil
+		imp.Audio = nil
+	case imp.Banner != nil:
+		imp.Video = nil
+		imp.Native = nil
+		imp.Audio = nil
+	default:
+	}
+
+	return nil
+}
+
+func assignBannerSize(banner *openrtb2.Banner) (*openrtb2.Banner, error) {
+	if banner.W != nil && banner.H != nil {
+		return banner, nil
+	}
+
+	if len(banner.Format) == 0 {
+		return banner, nil
+	}
+
+	return assignBannerWidthAndHeight(banner, banner.Format[0].W, banner.Format[0].H), nil
+}
+
+func assignBannerWidthAndHeight(banner *openrtb2.Banner, w, h int64) *openrtb2.Banner {
+	bannerCopy := *banner
+	bannerCopy.W = ptrutil.ToPtr(w)
+	bannerCopy.H = ptrutil.ToPtr(h)
+	return &bannerCopy
+}
+
+func getBidType(imps []openrtb2.Imp, impID string) openrtb_ext.BidType {
+	for _, imp := range imps {
+		if imp.ID != impID {
+			continue
+		}
+		switch {
+		case imp.Banner != nil:
+			return openrtb_ext.BidTypeBanner
+		case imp.Video != nil:
+			return openrtb_ext.BidTypeVideo
+		case imp.Native != nil:
+			return openrtb_ext.BidTypeNative
+		case imp.Audio != nil:
+			return openrtb_ext.BidTypeAudio
+		default:
+		}
+	}
+	return openrtb_ext.BidTypeBanner
```

instead of falling back to banner, should return error if no bid type matches.

---

moved to bid.mtype